### DISCUSSION TITLE
What is "can't connect, no HF2 nor JDUSB" and how to handle it

### DISCUSSION THREAD

**Discussion by civilizeddev** [microsoft/devicescript#447] [Q&A] [answered]
<img src="https://github.com/microsoft/jacdac-c/assets/15084045/0f870840-2fe4-4314-81dd-31e5d49b9af8" />

# What is "can't connect, no HF2 nor JDUSB" and how to handle it

Hello?

I have a question because I've been trying to do something with devicescript, but it's not working properly.

I failed to establish a serial connection to "ESP32-C3" board with an error message "can't connect, no HF2 nor JDUSB"

I'm not actually someone who works with devices professionally. I just wanted to try it out as a hobby. Instead of writing in C code, it seems that using TypeScript is an easier approach, so I wanted to give devicescript a try.

Therefore, I find it **difficult to understand the meaning of that message** and also **uncertain about what steps to take next to resolve the issue**.

An I also looked at the troubleshooting section in [this link](https://microsoft.github.io/devicescript/developer/errors#no-hf2), but I couldn't understand at all.

**If you could give me a rough idea of what steps I should take, I'll do my best to follow them.**

If there's any way I can contribute to resolving the issue, please let me know, and I'll be happy to help.

## Environment

I am a macOS user. (Ventura, Apple M1 Pro)

The board I have is "ESP32-C3-DevKitC-02" as shown in the image below. it belongs to the ESP32-C3 series.

<img src="https://github.com/microsoft/jacdac-c/assets/15084045/0adc276f-6b2d-485b-8855-8f22bd9ff44a" width="250" />

I can supply power through USB and establish serial communication by connecting to the UART, as shown below.

<img src="https://github.com/microsoft/jacdac-c/assets/15084045/6e96f365-b314-4c5d-81e6-4c6e2dcd8d35" width="500" />

Just in case it might work, I also tried connecting it as shown in the picture below.

<img src="https://github.com/microsoft/jacdac-c/assets/15084045/3afdb566-999c-45e8-9220-e1630034e820" width="500" />

## What I have tried

- I used the Arduino IDE to flash and run ESP32 example code on the board (It worked.)
- The board doesn't match exactly, but I selected `Espressif ESP32-C3 (bare)` and also tried flashing it.
  <img width="500" src="https://github.com/microsoft/jacdac-c/assets/15084045/03e5d8b5-43dc-4071-a518-240e2736d6c4"/>
  <img width="500" src="https://github.com/microsoft/jacdac-c/assets/15084045/2bbaed81-31b7-4407-9652-9b5b05f6111b" />
- However, when attempting to establish a serial connection, it doesn't work.
  <img width="500" src="https://github.com/microsoft/jacdac-c/assets/15084045/0f870840-2fe4-4314-81dd-31e5d49b9af8" />

## Related

- https://microsoft.github.io/devicescript/devices/esp32#cant-connect-no-hf2-nor-jdusb
- https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/get-started/establish-serial-connection.html

**Comment by pelikhan** [comment, marked as answer]
Unfortunately, we do not support ESP32-C3 boards with a serial interface chip. We need to have direct access to the serial line on the ESP32. You can see here a list of known boards that will work with DeviceScript. https://microsoft.github.io/devicescript/devices/esp32

**Reply by civilizeddev** [reply]
Is the chip limitation an issue that cannot be effectively addressed through devicescript (or jacdac)? Is it a current limitation or is there potential for improvement in the future? Thank you.

**Reply by pelikhan** [reply]
@mmoskal

**Reply by mmoskal** [reply]
The limitation is with the dev-board, not with the chip. It's unclear to me why are they putting an external USB serial chip on the official C3 dev-board, but luckily most other C3 dev boards connect the USB directly.

When we tried external USB on plain ESP32 it wasn't working very well when the Jacdac bus was noisy, so I don't expect us to support external USB on C3.

**Comment by pelikhan** [comment]
Thank you for this very detailled feedback. We've added a workitem to beef up the explanation about the interface chip limitations.
